staking/programs/staking/src/lib.rs

```diff
@@ -751,6 +754,157 @@ pub mod staking {
+
+                    // position_data.amount >= to_slash since slash_ratio is between 0 and 1
+                    if position_data.amount - to_slash == 0 {
+                        stake_account_positions.make_none(i, next_index)?;
+                        continue;
+                    } else {
+                        stake_account_positions.write_position(
+                            i,
+                            &Position {
+                                amount: position_data.amount - to_slash,
+                                target_with_parameters,
+                                activation_epoch: position_data.activation_epoch,
+                                unlocking_start: position_data.unlocking_start,
+                            },
+                        )?;
+                    }
+                }
+            }
+            i += 1;
+        }
+
+        let total_amount = ctx.accounts.stake_account_custody.amount;
+        let governance_exposure = calculate_governance_exposure(stake_account_positions)?;
+
+        let total_slashed = locked_slashed + unlocking_slashed + preunlocking_slashed;
+        if let Some(mut remaining) = (governance_exposure + total_slashed).checked_sub(total_amount)
+        {
+            let mut i = 0;
+            while i < usize::from(*next_index) && remaining > 0 {
+                if let Some(position) = stake_account_positions.read_position(i)? {
+                    let prev_state =
+                        position.get_current_position(current_epoch - 1, unlocking_duration)?;
+                    let current_state =
+                        position.get_current_position(current_epoch, unlocking_duration)?;
+
+                    if position.target_with_parameters == TargetWithParameters::Voting {
+                        let to_slash = remaining.min(position.amount);
+                        remaining -= to_slash;
+
+                        match prev_state {
+                            PositionState::LOCKED | PositionState::PREUNLOCKING => {
+                                governance_target_account
+                                    .sub_prev_locked(to_slash, current_epoch)?;
+                            }
+                            PositionState::LOCKING
+                            | PositionState::UNLOCKING
+                            | PositionState::UNLOCKED => {}
+                        }
+
+                        match current_state {
+                            PositionState::LOCKING => {
+                                governance_target_account.sub_locking(to_slash, current_epoch)?;
+                            }
+                            PositionState::LOCKED => {
+                                governance_target_account.sub_locked(to_slash, current_epoch)?;
+                            }
+                            PositionState::PREUNLOCKING => {
+                                governance_target_account.add_locking(to_slash, current_epoch)?;
```

Comment:
PREUNLOCKING also needs sub_locked if I'm not mistaken

Reply:
I thought when you initiate an unlock in epoch N, it will go to PREUNLOCKING in epoch n, as well as `add_unlocking` getting called. So if we want to revet this in epoch N we have to do `add_locking`, no?